srml/session/src/historical.rs

```diff
@@ -0,0 +1,425 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! An opt-in utility for tracking historical sessions in SRML-session.
+//!
+//! This is generally useful when implementing blockchains that require accountable
+//! safety where validators from some amount f prior sessions must remain slashable.
+//!
+//! Rather than store the full session data for any given session, we instead commit
+//! to the roots of merkle tries containing the session data.
+
+use rstd::prelude::*;
+use parity_codec::{Encode, Decode};
+use primitives::KeyTypeId;
+use primitives::traits::{Convert, OpaqueKeys, Hash as HashT};
+use srml_support::{
+	StorageValue, StorageMap, decl_module, decl_storage,
+};
+use srml_support::{Parameter, print};
+use substrate_trie::{MemoryDB, Trie, TrieMut, TrieDBMut, TrieDB, Recorder};
+
+use super::{SessionIndex, Module as SessionModule};
+
+/// Trait necessary for the historical module.
+pub trait Trait: super::Trait {
+	/// Full identification of the validator.
+	type FullIdentification: Parameter;
+
+	/// A conversion from validator ID to full identification.
+	///
+	/// This should contain any references to economic actors associated with the
+	/// validator, since they may be outdated by the time this is queried from a
+	/// historical trie.
+	///
+	/// This mapping is expected to remain stable in between calls to
+	/// `Self::OnSessionEnding::on_session_ending` which return new validators.
+	type FullIdentificationOf: Convert<Self::ValidatorId, Option<Self::FullIdentification>>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Session {
+		/// Mapping from historical session indices to session-data root hash.
+		HistoricalSessions get(historical_root): map SessionIndex => Option<T::Hash>;
+		/// Queued full identifications for queued sessions whose validators have become obsolete.
+		CachedObsolete get(cached_obsolete): map SessionIndex
```

Comment:
it should really only ever hold one or zero items, since we buffer at most 1 validator set in the current implementation. it only ever needs to hold something when we've done a validator election